{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import random\n",
    "\n",
    "# oh so pretty!\n",
    "def float_formatter(x):\n",
    "    if x < 0.001 and x > 0:\n",
    "        return \"   >0\" \n",
    "    elif x == 0:\n",
    "        return \"    0\" \n",
    "    else:\n",
    "        return \"%.3f\" % x\n",
    "np.set_printoptions(formatter={'float_kind':float_formatter})\n",
    "\n",
    "def _logsum(logx, logy):\n",
    "    print(\"Test\")\n",
    "    print(logx)\n",
    "    print(logy)\n",
    "    if len(logx.shape) < 1:\n",
    "        if np.isinf(logx) or np.isinf(logy):\n",
    "            print(\"result\")\n",
    "            print(np.max([logx, logy]))\n",
    "            return np.max([logx, logy])\n",
    "        else:\n",
    "            print(\"result\")\n",
    "            print(np.logaddexp(logx, logy))\n",
    "            return np.logaddexp(logx, logy)\n",
    "    else:\n",
    "        b = np.any([np.isinf(logx), np.isinf(logy)], axis=1)\n",
    "        result = np.zeros(shape=logx.shape[0])\n",
    "        result[b] = np.maximum(logx[b], logy[b])\n",
    "        result[not b] = np.logaddexp(logx[not b], logy[not b])\n",
    "        print(\"result\")\n",
    "        print(result)\n",
    "        return result\n",
    "\n",
    "def write_fasta_file(annotation_file_to_save, annotation_string, genome_file=None):\n",
    "    if (genome_file!=None):\n",
    "        with open(genome_file) as gf:\n",
    "            first_line = gf.readline()\n",
    "            s = \" \"\n",
    "            words = first_line.split(s)\n",
    "            words[0] = words[0] + \" gene annotation\"\n",
    "            first_line = s.join(words[:-1])\n",
    "    else:\n",
    "        first_line = \"; gene annotation\"\n",
    "    with open(annotation_file_to_save, 'w+') as af:\n",
    "        af.write(first_line + '\\n')\n",
    "        name = annotation_file_to_save.replace('/', '\\\\').split('\\\\')[-1].split('.fa')[0]\n",
    "        af.write('>' + name + '\\n')\n",
    "        line_length = 60\n",
    "        full_lines = len(annotation_string)//line_length\n",
    "        for anot_line in range(full_lines):\n",
    "            af.write(annotation_string[anot_line*line_length:(anot_line+1)*line_length] + '\\n')\n",
    "        af.write(annotation_string[(full_lines)*line_length:])\n",
    "       \n",
    "    \n",
    "# eg\n",
    "#write_fasta_file(\"data/annotation13.fa\", \"ABABABABABABABABABAAAAJHASJHASJJSdskfjklsfkjklj sdhjkøfhjkø ksdhjf jkjkhsdfjk hjksdf hjksdfjk jhksd fhjkdsfhjk AJJSJHASHJASJJHASJHASJHASJHSAJJHASJHASHJASJJHASJHASJJHASJH\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def read_fasta_file(filename):\n",
    "    \"\"\"\n",
    "    Reads the given FASTA file f and returns a dictionary of sequences.\n",
    "\n",
    "    Lines starting with ';' in the FASTA file are ignored.\n",
    "    \"\"\"\n",
    "    sequences_lines = {}\n",
    "    current_sequence_lines = None\n",
    "    with open(filename) as fp:\n",
    "        for line in fp:\n",
    "            line = line.strip()\n",
    "            if line.startswith(';') or not line:\n",
    "                continue\n",
    "            if line.startswith('>'):\n",
    "                sequence_name = line.lstrip('>')\n",
    "                current_sequence_lines = []\n",
    "                sequences_lines[sequence_name] = current_sequence_lines\n",
    "            else:\n",
    "                if current_sequence_lines is not None:\n",
    "                    current_sequence_lines.append(line)\n",
    "    sequences = {}\n",
    "    for name, lines in sequences_lines.items():\n",
    "        sequences[name] = ''.join(lines)\n",
    "    return sequences\n",
    "\n",
    "from collections import defaultdict\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "%matplotlib inline\n",
    "\n",
    "genome = ''\n",
    "annotation = ''\n",
    "for i in range(4,5):\n",
    "    genome += read_fasta_file('data/genome%d.fa' % i)['genome%d' % i]\n",
    "    annotation += read_fasta_file('data/annotation%d.fa' % i)['annotation%d' % i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1.000     0     0]\n",
      " [    0     0 1.000]\n",
      " [    0     0 1.000]\n",
      " [    0     0 1.000]\n",
      " [    0     0 1.000]\n",
      " [    0     0 1.000]\n",
      " [    0     0 1.000]]\n",
      "[1.000     0     0     0     0     0     0]\n",
      "['Noncoding' 'C-start' 'C-codon' 'C-stop' 'R-start' 'R-codon' 'R-stop']\n",
      "['N' 'C' 'C' 'C' 'R' 'R' 'R']\n"
     ]
    }
   ],
   "source": [
    "#Here we load our model\n",
    "npzfile = np.load('codon_hmm_excl5 - Copy.npz')\n",
    "A = npzfile['A']\n",
    "E = npzfile['E']\n",
    "pi = npzfile['pi']\n",
    "d = np.zeros((A.shape[0], 3))\n",
    "d[0, 0] = 1\n",
    "for i in range(1, d.shape[0]):\n",
    "    d[i, 2] = 1\n",
    "print(d)\n",
    "print(d[:, 0])\n",
    "\n",
    "# define all possible outcomes of all states:\n",
    "outcomes_dict = {} # ohh gawd this is akward\n",
    "outcomes = np.empty([68], dtype=object) # help me, im lost\n",
    "i = 0\n",
    "letters = 'ACGT'\n",
    "for letter1 in letters:\n",
    "    outcomes[i] = letter1\n",
    "    outcomes_dict[letter1] = i\n",
    "    i += 1\n",
    "    for letter2 in letters:\n",
    "        for letter3 in letters:\n",
    "            outcomes[i] = letter1 + letter2 + letter3\n",
    "            outcomes_dict[letter1 + letter2 + letter3] = i\n",
    "            i += 1\n",
    "            \n",
    "\n",
    "\n",
    "state_codes=npzfile['state_codes']\n",
    "state_desc=npzfile['state_desc']\n",
    "\n",
    "q = 3*np.ones(len(state_desc), dtype=np.int32)\n",
    "q[0] = 1\n",
    "print(state_desc)\n",
    "print(state_codes)\n",
    "\n",
    "def l2i(l):\n",
    "    if type(l) is str:\n",
    "        if l == \"\" or len(l) == 2:\n",
    "            return -1\n",
    "        else:\n",
    "            return outcomes_dict[l]\n",
    "    else:\n",
    "        return [l2i(l_) for l_ in l]\n",
    "#i can either be a list of indices, or just an index\n",
    "def i2l(i):\n",
    "    if type(i) is int:\n",
    "        return outcomes[i]\n",
    "    else:\n",
    "        return [outcomes[i_] for i_ in i]\n",
    "def e_prop(E, s):\n",
    "    result = list()\n",
    "    for i in range(len(s)):\n",
    "        result.append(E[i, s[i]])\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Morten\\Anaconda3\\lib\\site-packages\\ipykernel\\__main__.py:14: RuntimeWarning: divide by zero encountered in log\n",
      "C:\\Users\\Morten\\Anaconda3\\lib\\site-packages\\ipykernel\\__main__.py:15: RuntimeWarning: divide by zero encountered in log\n",
      "C:\\Users\\Morten\\Anaconda3\\lib\\site-packages\\ipykernel\\__main__.py:16: RuntimeWarning: divide by zero encountered in log\n",
      "C:\\Users\\Morten\\Anaconda3\\lib\\site-packages\\ipykernel\\__main__.py:17: RuntimeWarning: divide by zero encountered in log\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[    0 -inf -inf -inf -inf -inf -inf]\n",
      "[    0 -inf -inf -inf -inf -inf -inf]\n",
      "[-inf -inf -3.415 -inf -inf -2.899 -inf]\n",
      "[[-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " ..., \n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]]\n",
      "\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "[[-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " ..., \n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]]\n",
      "\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "Test\n",
      "-inf\n",
      "-inf\n",
      "result\n",
      "-inf\n",
      "[[-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " ..., \n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]\n",
      " [-inf -inf -inf ..., -inf -inf -inf]]\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Morten\\Anaconda3\\lib\\site-packages\\ipykernel\\__main__.py:19: RuntimeWarning: divide by zero encountered in log\n"
     ]
    }
   ],
   "source": [
    "#This is the Viterbi algorithm, where different states, can take input of different length.\n",
    "#A, E, p and s is as above. q is a vector, signifying the length of inputs for each state.\n",
    "\n",
    "s = genome\n",
    "\n",
    "def logaddexplist(l):\n",
    "    result = 0\n",
    "    for element in l:\n",
    "        result = np.logaddexp(result, element)\n",
    "    return result\n",
    "\n",
    "\n",
    "def viterbi_ext_gen(A, E, p, s, q, d):\n",
    "    A = np.log(A)\n",
    "    E = np.log(E)\n",
    "    p = np.log(p)\n",
    "    d = np.log(d)\n",
    "    #Then we generate some result matrix, with -inf everywhere.\n",
    "    result = np.log(np.zeros([A.shape[0], len(s)]))\n",
    "    \n",
    "    #First we make the base cases\n",
    "    obs = l2i([s[0:i] for i in range(d.shape[1])])\n",
    "    #first row\n",
    "    result[:, 0] = p[:,0] + d[:, 0] + E[:, obs[0]]\n",
    "    print(p[:,0])\n",
    "    print(d[:, 0])\n",
    "    print(E[:, obs[0]])\n",
    "    print(result)\n",
    "    print()\n",
    "    #second row\n",
    "    result[:, 1] = p[:,0] + d[:, 1] + np.sum(E[:, [obs[0], obs[1]]], axis=1)\n",
    "    for i in range(A.shape[0]):\n",
    "        for j in range(A.shape[0]):\n",
    "            if j != i:\n",
    "                result[i, 1] = _logsum(result[i, 1], result[j, 0] + A[i,j] + d[i, 1] + E[i, obs[1]])\n",
    "                                  \n",
    "    result[:, 2] = p[:,0] + d[:, 2] + np.sum(E[:, [obs[0], obs[1], obs[2]]], axis=1)\n",
    "    print(result)\n",
    "    print()\n",
    "    for k in range(1, 3):\n",
    "        for i in range(A.shape[0]):\n",
    "            for j in range(A.shape[0]):\n",
    "                if j != i:\n",
    "                    e_ = 0\n",
    "                    for l in range(3-k, 3):\n",
    "                        e_ += E[i, obs[l]]\n",
    "                    result[i, 2] = _logsum(result[i, 2], result[j, 3-k] + A[i,j] + d[i, 2] + e_)\n",
    "    print(result)\n",
    "    print()\n",
    "    return result\n",
    "\n",
    "def viterbi_ext_path(X, A, E, p, s, q):\n",
    "    A = np.log(A)\n",
    "    E = np.log(E)\n",
    "    p = np.log(p)\n",
    "    n = X.shape[1]-2\n",
    "    j = len(s)-2\n",
    "    path = -1*np.ones(n+1, dtype=np.int32)\n",
    "    path[-2::] = 0\n",
    "    for i in range(n, 0, -1):\n",
    "        q_ = q[path[i]]\n",
    "        s_ = s[j-q_:j+(3-q_)]\n",
    "        p = E[path[i], l2i(s_)]\n",
    "        p = p + X[:, i-1]\n",
    "        p = p + A[:, path[i]]\n",
    "        path[i-1] = np.argmax(p)\n",
    "        j = j-q_\n",
    "    return path\n",
    "\n",
    "def viterbi_ext_annotation(genome, path):\n",
    "    result = np.empty(len(path), dtype=object)\n",
    "    result[path == 0] = 'N'\n",
    "    result[path == 1] = 'CCC'\n",
    "    result[path == 2] = 'CCC'\n",
    "    result[path == 3] = 'CCC'\n",
    "    result[path == 4] = 'RRR'\n",
    "    result[path == 5] = 'RRR'\n",
    "    result[path == 6] = 'RRR'\n",
    "    print(result.shape)\n",
    "    result = \"\".join(result)\n",
    "    print(len(result))\n",
    "    return result\n",
    "\n",
    "X = viterbi_ext_gen(A, E, pi, s, q, d)\n",
    "#print(X)\n",
    "#path = viterbi_ext_path(X, A, E, pi, s, q)\n",
    "#print(path)\n",
    "#print((path!=0).sum())\n",
    "#for i in range(path.shape[0]):\n",
    "#    if path[i] == 1:\n",
    "#        print(X[:, i-2:i+2])\n",
    "#        break\n",
    "#annotation_pred = viterbi_ext_annotation(s, path)\n",
    "#print(len(genome))\n",
    "#print(len(annotation_pred))\n",
    "#print(annotation_pred)\n",
    "#write_fasta_file('data/annotation_prediction_5.fa', annotation, 'data/genome5.fa')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0 1]\n",
      " [2 3]]\n",
      "[0 6]\n"
     ]
    }
   ],
   "source": [
    "S = np.arange(4).reshape(2,2)\n",
    "print(S)\n",
    "print(S.prod(axis=1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR: The lengths of two predictions are different\n",
      "Expected 1796846, but found 2053385\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "1",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\u001b[0;31m:\u001b[0m 1\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/morten/anaconda3/lib/python3.5/site-packages/IPython/core/interactiveshell.py:2889: UserWarning: To exit: use 'exit', 'quit', or Ctrl-D.\n",
      "  warn(\"To exit: use 'exit', 'quit', or Ctrl-D.\", stacklevel=1)\n"
     ]
    }
   ],
   "source": [
    "#\n",
    "# compare_anns.py <true> <pred>\n",
    "#\n",
    "# compares a predicted gene structure against the true gene structure and computes\n",
    "# various statistics summarizing the quality of the prediction. The argument <true> is \n",
    "# the true gene structure in faste format, and <pred> is the predicted gene structure \n",
    "# in fasta format, e.g.\n",
    "#\n",
    "# > python compare_anns.py ./annotation1.fa ./pred1.fa\n",
    "# > Only Cs (tp=728238, fp=0, tn=505177, fn=249):\n",
    "# > Sn = 0.9997, Sp = 1.0000, CC = 0.9996, AC = 0.9996\n",
    "# > Only Rs (tp=618777, fp=0, tn=505426, fn=0):\n",
    "# > Sn = 1.0000, Sp = 1.0000, CC = 1.0000, AC = 1.0000\n",
    "# > Both (tp=1347015, fp=0, tn=505177, fn=249):\n",
    "# > Sn = 0.9998, Sp = 1.0000, CC = 0.9997, AC = 0.9997\n",
    "#\n",
    "# Christian Storm <cstorm@birc.au.dk>\n",
    "#\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import string\n",
    "import math\n",
    "\n",
    "def read_ann(filename):\n",
    "    lines = []\n",
    "    for l in open(filename).readlines():\n",
    "        if l[0] != \">\" and l[0] != ';':\n",
    "            lines.append(l.strip())\n",
    "    return \"\".join(lines)\n",
    "\n",
    "def count_c(true, pred):\n",
    "    total = tp = fp = tn = fn = 0\n",
    "    for i in range(len(true)):\n",
    "        if pred[i] == 'C' or pred[i] == 'c':\n",
    "            total = total + 1\n",
    "            if true[i] == 'C' or true[i] == 'c':\n",
    "                tp = tp + 1\n",
    "            else:\n",
    "                fp = fp + 1\n",
    "        if pred[i] == 'N' or pred[i] == 'n':\n",
    "            if true[i] == 'N' or true[i] == 'n' or true[i] == 'R' or true[i] == 'r':\n",
    "                tn = tn + 1\n",
    "            else:\n",
    "                fn = fn + 1\n",
    "    return(total, tp, fp, tn, fn)\n",
    "\n",
    "def count_r(true, pred):\n",
    "    total = tp = fp = tn = fn = 0\n",
    "    for i in range(len(true)):\n",
    "        if pred[i] == 'R' or pred[i] == 'r':\n",
    "            total = total + 1\n",
    "            if true[i] == 'R' or true[i] == 'r':\n",
    "                tp = tp + 1\n",
    "            else:\n",
    "                fp = fp + 1\n",
    "        if pred[i] == 'N' or pred[i] == 'n':\n",
    "            if true[i] == 'N' or true[i] == 'n' or true[i] == 'C' or true[i] == 'c':\n",
    "                tn = tn + 1\n",
    "            else:\n",
    "                fn = fn + 1\n",
    "    return(total, tp, fp, tn, fn)\n",
    "\n",
    "def count_cr(true, pred):\n",
    "    total = tp = fp = tn = fn = 0\n",
    "    for i in range(len(true)):\n",
    "        if pred[i] == 'C' or pred[i] == 'c' or pred[i] == 'R' or pred[i] == 'r':\n",
    "            total = total + 1\n",
    "            if (pred[i] == 'C' or pred[i] == 'c') and (true[i] == 'C' or true[i] == 'c'):\n",
    "                tp = tp + 1\n",
    "            elif (pred[i] == 'R' or pred[i] == 'r') and (true[i] == 'R' or true[i] == 'r'):\n",
    "                tp = tp + 1                \n",
    "            else:\n",
    "                fp = fp + 1\n",
    "        if pred[i] == 'N' or pred[i] == 'n':\n",
    "            if true[i] == 'N' or true[i] == 'n':\n",
    "                tn = tn + 1\n",
    "            else:\n",
    "                fn = fn + 1\n",
    "    return(total, tp, fp, tn, fn)\n",
    "\n",
    "def print_stats(tp, fp, tn, fn):\n",
    "    sn = float(tp) / (tp + fn)\n",
    "    sp = float(tp) / (tp + fp)\n",
    "    cc = float((tp*tn - fp*fn)) / math.sqrt(float((tp+fn)*(tn+fp)*(tp+fp)*(tn+fn)))\n",
    "    acp = 0.25 * (float(tp)/(tp+fn) + float(tp)/(tp+fp) + float(tn)/(tn+fp) + float(tn)/(tn+fn))\n",
    "    ac = (acp - 0.5) * 2\n",
    "    print(\"Sn = %.4f, Sp = %.4f, CC = %.4f, AC = %.4f\" % (sn, sp, cc, ac))\n",
    "\n",
    "def print_all(true, pred):\n",
    "    (totalc, tp, fp, tn, fn) = count_c(true, pred)\n",
    "    if totalc > 0:\n",
    "        print(\"Only Cs (tp=%d, fp=%d, tn=%d, fn=%d):\" % (tp, fp, tn, fn))\n",
    "        print_stats(tp, fp, tn, fn)\n",
    "\n",
    "    (totalr, tp, fp, tn, fn) = count_r(true, pred)\n",
    "    if totalr > 0:\n",
    "        print(\"Only Rs (tp=%d, fp=%d, tn=%d, fn=%d):\" % (tp, fp, tn, fn))\n",
    "        print_stats(tp, fp, tn, fn)\n",
    "\n",
    "    (total, tp, fp, tn, fn) = count_cr(true, pred)\n",
    "    if totalc > 0 and totalr > 0:\n",
    "        print(\"Both (tp=%d, fp=%d, tn=%d, fn=%d):\" % (tp, fp, tn, fn))\n",
    "        print_stats(tp, fp, tn, fn)\n",
    "\n",
    "\n",
    "# Read true annotation\n",
    "true_ann = read_ann('data/annotation5.fa')\n",
    "\n",
    "# Read predicted annotations\n",
    "pred_ann = read_ann('data/annotation_prediction_5.fa')\n",
    "# Check annoation length\n",
    "error = 0\n",
    "if len(true_ann) != len(pred_ann):\n",
    "    print(\"ERROR: The lengths of two predictions are different\")\n",
    "    print(\"Expected %d, but found %d\" % (len(true_ann), len(pred_ann)))\n",
    "    sys.exit(1)    \n",
    "    \n",
    "# Print stats\n",
    "print_all(true_ann, pred_ann)\n"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
