
\documentclass[a4paper,10pt,article,oneside,english]{memoir} 
% DANSK OPSÃ†TNING
\usepackage[english]{babel}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{lmodern} 
\renewcommand{\englishhyphenmins}{22} 


% FIGURER
\usepackage{graphicx}

% MATEMATIK
\usepackage{amsmath}
\usepackage{amssymb,amsthm,bm}
\usepackage{mathtools}	


% This document uses
\usepackage[draft,silent]{fixme}
\usepackage{hyperref}
\usepackage{siunitx}
\usepackage{booktabs}

% captions in italic
\let\oldcaption\caption
\renewcommand{\caption}[1]{\oldcaption{\emph{#1}}}


\begin{document}
	%\frontmatter
	%\clearpage	
	%\tableofcontents*
	\title{Machine Learning E16 - Handin 3\\
		 Hidden Markov Model for Gene Finding in Prokaryotes}
	\author{Group 22\\
		Mark Medum Bundgaard, Morten Jensen, Martin Sand Nielsen}
	\date{\today, Aarhus University}
	
	\mainmatter
	\maketitle

\section{Model}

\begin{figure}
	\centering
	\includegraphics[width=\linewidth]{HMM_graph_cropped.pdf}
	\caption{Transition diagram and state emissions for our HMM for genome annotation by explicitly matching codons within coding areas, and enforcing certain start and stop codons. Transition and emission probabilities approximated by counting annotated data.}
	\label{fig:hmm_graph}
\end{figure}
	
\begin{table}
	\centering
	\caption{text}
	\label{tab:ac}
	\begin{tabular}{rccc}
		& \multicolumn{3}{c}{Approximate correlation coefficient} \\ 
		validation & C & R & Both \\ 
		\midrule
		Genome 1 &  &  &  \\ 
		Genome 2 &  &  &  \\ 
		Genome 3 &  &  &  \\ 
		Genome 4 &  &  &  \\ 
		Genome 5 &  &  &  \\ 
		Average & & & \\
	\end{tabular} 
\end{table}
	
\end{document}
